### PR TITLE
Add a CupertinoTextField

### PR DESCRIPTION
**Pull request by xster**
Adds support for an iOS styled text field #13695 with APIs supporting default UITextField and UITextView functionalities. 

Consolidates docs between Material TextField, EditableText and this new widget.

### CODE REVIEW

File: packages/flutter/lib/src/cupertino/app.dart [hunk 1]
```diff
@@ -243,7 +243,7 @@ class _CupertinoAppState extends State<CupertinoApp> {
       _navigatorObservers = List<NavigatorObserver>.from(widget.navigatorObservers)
         ..add(_heroController);
     } else {
-      _navigatorObservers = null;
+      _navigatorObservers = const <NavigatorObserver>[];
```
Comment (matthew-carroll) [inline packages/flutter/lib/src/cupertino/app.dart]: is this an unrelated change?

File: packages/flutter/lib/src/cupertino/icons.dart [hunk 2]
```diff
@@ -721,4 +721,44 @@ class CupertinoIcons {
   ///  * [group], which is similar, but not filled in.
   ///  * [person_solid], which is just a single person.
   static const IconData group_solid = IconData(0xf47c, fontFamily: iconFont, fontPackage: iconFontPackage);
+
+  /// Outline of a closed mail envelope.
+  static const IconData mail = IconData(0xf422, fontFamily: iconFont, fontPackage: iconFontPackage);
```
Comment (matthew-carroll) [inline packages/flutter/lib/src/cupertino/icons.dart]: Glad to see more icons added :)

File: packages/flutter/lib/src/cupertino/text_field.dart [hunk 3]
```diff
@@ -0,0 +1,600 @@
...
+);
+const BoxDecoration _kDefaultRoundedBorderDecoration = BoxDecoration(
+  border: _kDefaultRoundedBorder,
+  borderRadius: BorderRadius.all(Radius.circular(4.0)),
+);
+
+const TextStyle _kDefaultTextStyle = TextStyle(
+  fontFamily: '.SF Pro Text',
+  fontSize: 17.0,
+  letterSpacing: -0.38,
```
Comment (matthew-carroll) [inline packages/flutter/lib/src/cupertino/text_field.dart]: Out of curiosity, where did the letter spacing come from?

File: packages/flutter/lib/src/cupertino/text_field.dart [hunk 4]
```diff
@@ -0,0 +1,600 @@
...
+  color: CupertinoColors.black,
+  decoration: TextDecoration.none,
+);
+
+const Color _kSelectionHighlightColor = Color(0x667FAACF);
```
Comment (matthew-carroll) [inline packages/flutter/lib/src/cupertino/text_field.dart]: A thought on all of these constants we have throughout cupertino files...I'm thinking that maybe we should aggregate all such constants in a single cupertino file so that other developers can use them, if desired.  I'm sure developers will need to create their own iOS styled widgets and it would be nice if they had a way to ensure they're using the same colors, border radii, etc. that we use.

File: packages/flutter/lib/src/cupertino/text_field.dart [hunk 5]
```diff
@@ -0,0 +1,600 @@
...
+const Color _kInactiveTextColor = Color(0xFFC2C2C2);
+const Color _kDisabledBackground = Color(0xFFFAFAFA);
+
+/// Visibility of text field overlays based on the state of the current text entry.
```
Comment (matthew-carroll) [inline packages/flutter/lib/src/cupertino/text_field.dart]: Can you describe what a "text field overlay" is?

File: packages/flutter/lib/src/cupertino/text_field.dart [hunk 6]
```diff
@@ -0,0 +1,600 @@
...
+  /// This also includes not having pre-filled text that the user did not type
+  /// in manually. Texts in placeholders are ignored.
+  notEditing,
+  /// Always show the overlay regardless of the text entry state.
+  always,
+}
+
+/// An iOS-style text field.
+///
+/// A text field lets the user enter text, either with hardware keyboard or with
```
Comment (matthew-carroll) [inline packages/flutter/lib/src/cupertino/text_field.dart]: typo: with *a* hardware keyboard

File: packages/flutter/lib/src/cupertino/text_field.dart [hunk 7]
```diff
@@ -0,0 +1,600 @@
...
+/// an onscreen keyboard.
+///
+/// This widget corresponds to both a [UITextField] and an editable [UITextView]
```
Comment (matthew-carroll) [inline packages/flutter/lib/src/cupertino/text_field.dart]: Is there a reason that we want to combine these 2 concepts?  Does a UITextView every become a UITextField?  If not, then we might consider wrapping the common implementation with 2 public widgets instead of 1.

File: packages/flutter/lib/src/cupertino/text_field.dart [hunk 8]
```diff
@@ -0,0 +1,600 @@
...
+/// on iOS.
+///
+/// The text field calls the [onChanged] callback whenever the user changes the
+/// text in the field. If the user indicates that they are done typing in the
+/// field (e.g., by pressing a button on the soft keyboard), the text field
+/// calls the [onSubmitted] callback.
+///
+/// To control the text that is displayed in the text field, use the
```
Comment (matthew-carroll) [inline packages/flutter/lib/src/cupertino/text_field.dart]: I think a couple of code snippets here would probably be much more clear than a description of what to do.  You can show a code snippet for setting an initial value, and a code snippet for selecting text.

File: packages/flutter/lib/src/cupertino/text_field.dart [hunk 9]
```diff
@@ -0,0 +1,600 @@
...
+/// [controller]. For example, to set the initial value of the text field, use
+/// a [controller] that already contains some text. The [controller] can also
+/// control the selection and composing region (and to observe changes to the
+/// text, selection, and composing region).
+///
+/// The text field has an overridable [decoration] that, by default, draws a
```
Comment (matthew-carroll) [inline packages/flutter/lib/src/cupertino/text_field.dart]: Do we want people to be able to control this?  Does a UITextField ever not have the decoration in normal iOS development?

File: packages/flutter/lib/src/cupertino/text_field.dart [hunk 10]
```diff
@@ -0,0 +1,600 @@
...
+    assert(debugCheckHasDirectionality(context));
+    final Brightness keyboardAppearance = widget.keyboardAppearance;
+    final TextEditingController controller = _effectiveController;
+    final FocusNode focusNode = _effectiveFocusNode;
+    final List<TextInputFormatter> formatters = widget.inputFormatters ?? <TextInputFormatter>[];
+    final bool enabled = widget.enabled ?? true;
+    if (widget.maxLength != null && widget.maxLengthEnforced)
+      formatters.add(LengthLimitingTextInputFormatter(widget.maxLength));
+
+    Widget child = Padding(
```
Comment (matthew-carroll) [inline packages/flutter/lib/src/cupertino/text_field.dart]: Given the proliferation of the term "child" in widget trees, can you name this variable based on what it represents within the CupertinoTextField UI?

File: packages/flutter/lib/src/cupertino/text_field.dart [hunk 11]
```diff
@@ -0,0 +1,600 @@
...
+          cursorWidth: widget.cursorWidth,
+          cursorRadius: widget.cursorRadius,
+          cursorColor: widget.cursorColor,
+          scrollPadding: widget.scrollPadding,
+          keyboardAppearance: keyboardAppearance,
+        ),
+      ),
+    );
+
+    if (widget.placeholder != null ||
```
Comment (matthew-carroll) [inline packages/flutter/lib/src/cupertino/text_field.dart]: Can we get some comments throughout this code explaining what the conditions mean and what's being done about it?

File: packages/flutter/lib/src/cupertino/text_field.dart [hunk 12]
```diff
@@ -0,0 +1,600 @@
...
+        widget.clearButtonMode != OverlayVisibilityMode.never ||
+        widget.prefix != null ||
+        widget.suffix != null) {
+      child = ValueListenableBuilder<TextEditingValue>(
```
Comment (matthew-carroll) [inline packages/flutter/lib/src/cupertino/text_field.dart]: I think re-assignment of 'child' is probably unnecessarily confusing for readers.  Can we stick to immutable values as we compose the tree?

File: packages/flutter/lib/src/cupertino/text_field.dart [hunk 13]
```diff
@@ -0,0 +1,600 @@
...
+        valueListenable: controller,
+        builder: (BuildContext context, TextEditingValue text, Widget child) {
+          final List<Widget> rowChildren = <Widget>[];
+
+          if (widget.prefix != null &&
+              _shouldAppearForModeAndText(widget.prefixMode, text.text.isNotEmpty)) {
+            rowChildren.add(widget.prefix);
```
Comment (matthew-carroll) [inline packages/flutter/lib/src/cupertino/text_field.dart]: Is there any natural spacing around the prefix?  I think we should ensure that a developer who simply drops an Icon in as the prefix gets the expected spacing that appears in a real iOS UITextField

File: packages/flutter/lib/src/cupertino/text_field.dart [hunk 14]
```diff
@@ -0,0 +1,600 @@
...
+        widget.clearButtonMode != OverlayVisibilityMode.never ||
+        widget.prefix != null ||
+        widget.suffix != null) {
+      child = ValueListenableBuilder<TextEditingValue>(
+        valueListenable: controller,
+        builder: (BuildContext context, TextEditingValue text, Widget child) {
+          final List<Widget> rowChildren = <Widget>[];
+
+          if (widget.prefix != null &&
+              _shouldAppearForModeAndText(widget.prefixMode, text.text.isNotEmpty)) {
```
Comment (matthew-carroll) [inline packages/flutter/lib/src/cupertino/text_field.dart]: This method name is strange for this class.  Based on the behavior it looks like it should be a method of a prefix, for example: prefix.shouldAppearForModeAndText(...).  If it's going to be defined on this class, can you add a reference to 'prefix' so that it's clear what "should appear" or not?

File: packages/flutter/lib/src/cupertino/text_field.dart [hunk 15]
```diff
@@ -0,0 +1,600 @@
...
+
+          final List<Widget> stackChildren = <Widget>[];
+
+          if (widget.placeholder != null && text.text.isEmpty) {
+            stackChildren.add(
+              Padding(
+                padding: widget.padding,
+                child: Text(
+                  widget.placeholder,
+                  maxLines: 1,
```
Comment (matthew-carroll) [inline packages/flutter/lib/src/cupertino/text_field.dart]: Should there be an overflow specified here?

File: packages/flutter/lib/src/cupertino/text_field.dart [hunk 16]
```diff
@@ -0,0 +1,600 @@
...
+                  ),
+                ),
+              ),
+            );
+          }
+
+          rowChildren.add(Expanded(child: Stack(children: stackChildren..add(child))));
+
+          if (widget.suffix != null &&
+              _shouldAppearForModeAndText(widget.suffixMode, text.text.isNotEmpty)) {
```
Comment (matthew-carroll) [inline packages/flutter/lib/src/cupertino/text_field.dart]: Ah, I see, this method takes in many different things.  I would definitely break this method apart.  As a reader, I'd like to fully understand the nature of what it is being called, but methods that essentially take in switchable values require a developer to jump into the method to understand what it's really used for.

So if you want to ask the question "should we show the prefix", then I'd like to see a method like _shouldShowPrefix(...), and then _shoudlShowSuffix(...).  That greatly reduces the cognitive overhead of reading through the code.

File: packages/flutter/lib/src/cupertino/text_field.dart [hunk 17]
```diff
@@ -0,0 +1,600 @@
...
+            rowChildren.add(widget.suffix);
+          } else if (_shouldAppearForModeAndText(widget.clearButtonMode, text.text.isNotEmpty)) {
```
Comment (matthew-carroll) [inline packages/flutter/lib/src/cupertino/text_field.dart]: Here again with the hiding of the specific decision being made.

File: packages/flutter/lib/src/cupertino/text_field.dart [hunk 18]
```diff
@@ -0,0 +1,600 @@
...
+            rowChildren.add(
+              GestureDetector(
+                onTap: enabled
+                    ? () => controller.clear()
+                    : null,
+                child: const Padding(
+                  padding: EdgeInsets.symmetric(horizontal: 6.0),
+                  child: Icon(
+                    CupertinoIcons.clear_thick_circled,
```
Comment (matthew-carroll) [inline packages/flutter/lib/src/cupertino/text_field.dart]: I'll ask a similar question here as elsewhere, are we trying to allow customization or not?  If we are, then I think this needs to be customizable.  Otherwise, if we want to enforce true iOS styling of this widget then we should remove customization controls.

File: packages/flutter/lib/src/cupertino/text_field.dart [hunk 19]
```diff
@@ -0,0 +1,600 @@
...
+
+          return Row(children: rowChildren);
+        },
+        child: child,
+      );
+    }
+
+    return Semantics(
+      onTap: () {
+        if (!_effectiveController.selection.isValid)
```
Comment (matthew-carroll) [inline packages/flutter/lib/src/cupertino/text_field.dart]: Would you mind wrapping this if-statement with braces?  It is especially easy to misread an implied single-line if-statement when it's sitting within a larger structure like this.

**Comment by xster** [comment]
Thanks for the quick review!

File: packages/flutter/lib/src/cupertino/app.dart (hunk 1 above)
Comment (xster) [reply packages/flutter/lib/src/cupertino/app.dart]: Fixes a bug the demo revealed. The observer list can't be null.

File: packages/flutter/lib/src/cupertino/text_field.dart (hunk 3 above)
Comment (xster) [reply packages/flutter/lib/src/cupertino/text_field.dart]: Good point. Adding comments in code about their provenance.

File: packages/flutter/lib/src/cupertino/text_field.dart (hunk 4 above)
Comment (xster) [reply packages/flutter/lib/src/cupertino/text_field.dart]: I agree. Common ones should be in cupertino/colors.dart. Though iOS doesn't really semantically catalog them in a public place so I hesitate to put new ones in there each time I encounter a new color since namespace real estate is valuable. I generally comb the code each time I see a color and if I can figure out a pattern, I add it to CupertinoColors. For the ones that are seemingly domain specific like these ones as far as I can tell for now, I leave them private in their place of use.

File: packages/flutter/lib/src/cupertino/text_field.dart (hunk 5 above)
Comment (xster) [reply packages/flutter/lib/src/cupertino/text_field.dart]: Adding more text in a next paragraph

File: packages/flutter/lib/src/cupertino/text_field.dart (hunk 6 above)
Comment (xster) [reply packages/flutter/lib/src/cupertino/text_field.dart]: thanks

File: packages/flutter/lib/src/cupertino/text_field.dart (hunk 7 above)
Comment (xster) [reply packages/flutter/lib/src/cupertino/text_field.dart]: We should make all the default platform APIs available and easy to use but not necessarily mirror the APIs. In this case, the iOS APIs aren't great examples. There's tons of duplications between the 2 and the only real difference is one being single-lined. It would be an annoying user experience to split them into 2.

File: packages/flutter/lib/src/cupertino/text_field.dart (hunk 8 above)
Comment (xster) [reply packages/flutter/lib/src/cupertino/text_field.dart]: Good point. Adding

File: packages/flutter/lib/src/cupertino/text_field.dart (hunk 9 above)
Comment (xster) [reply packages/flutter/lib/src/cupertino/text_field.dart]: Yes, the iOS text fields are fairly customizable

File: packages/flutter/lib/src/cupertino/text_field.dart (hunk 10 above)
Comment (xster) [reply packages/flutter/lib/src/cupertino/text_field.dart]: I'm breaking this build up a bit

File: packages/flutter/lib/src/cupertino/text_field.dart (hunk 12 above)
Comment (xster) [reply packages/flutter/lib/src/cupertino/text_field.dart]: Refactored

File: packages/flutter/lib/src/cupertino/text_field.dart (hunk 11 above)
Comment (xster) [reply packages/flutter/lib/src/cupertino/text_field.dart]: Adding comments

File: packages/flutter/lib/src/cupertino/text_field.dart (hunk 13 above)
Comment (xster) [reply packages/flutter/lib/src/cupertino/text_field.dart]: Agree. Though it turned out fairly comparable to iOS without more padding here.

File: packages/flutter/lib/src/cupertino/text_field.dart (hunk 14 above)
Comment (xster) [reply packages/flutter/lib/src/cupertino/text_field.dart]: Ya, reversing it would be nice but the prefix mode is an enum and dart enums aren't like classes

File: packages/flutter/lib/src/cupertino/text_field.dart (hunk 15 above)
Comment (xster) [reply packages/flutter/lib/src/cupertino/text_field.dart]: good call! added

File: packages/flutter/lib/src/cupertino/text_field.dart (hunk 18 above)
Comment (xster) [reply packages/flutter/lib/src/cupertino/text_field.dart]: Is this in reference to the clear button? It's just matching the flexibility of the iOS APIs. iOS's got one quick way of turning on a fixed looking clear button on text fields. Otherwise, it expands quickly from there and any UIView can go into the left and right view slots.

File: packages/flutter/lib/src/cupertino/text_field.dart (hunk 19 above)
Comment (xster) [reply packages/flutter/lib/src/cupertino/text_field.dart]: yup. I dislike the optional braces thing too

File: packages/flutter/lib/src/cupertino/text_field.dart (hunk 16 above)
Comment (xster) [reply packages/flutter/lib/src/cupertino/text_field.dart]: Ya, I don't have a super great solution. Though more indirections impedes readability too. Hopefully the compromise is reasonable by shortening the method and highlighting the key part more quickly with named params.

File: packages/flutter/lib/src/cupertino/text_field.dart [hunk 20]
```diff
@@ -0,0 +1,657 @@
...
+  /// This includes pre-filled text that the user did not type in manually. But
+  /// does not include text in placeholders.
+  editing,
+  /// Overlay will only appear when the current text entry is empty.
+  ///
+  /// This also includes not having pre-filled text that the user did not type
+  /// in manually. Texts in placeholders are ignored.
+  notEditing,
+  /// Always show the overlay regardless of the text entry state.
+  always,
```
Comment (Hixie) [inline packages/flutter/lib/src/cupertino/text_field.dart]: nit: blank line between an enum value and the next's dartdoc

File: packages/flutter/lib/src/cupertino/text_field.dart (hunk 7 above)
Comment (Hixie) [reply packages/flutter/lib/src/cupertino/text_field.dart]: UITextField and the other one should be in backticks, not square brackets

File: packages/flutter/lib/src/cupertino/text_field.dart [hunk 21]
```diff
@@ -0,0 +1,657 @@
...
+///     super.initState();
+///     _textController = TextEditingController(text: 'initial text');
+///   }
+///
+///   @override
+///   Widget build(BuildContext context) {
+///     return CupertinoTextField(controller: _textController);
+///   }
+/// }
+/// ```
```
Comment (Hixie) [inline packages/flutter/lib/src/cupertino/text_field.dart]: sample code should be in a `## Sample code` section so it gets analyzed.

File: packages/flutter/lib/src/cupertino/text_field.dart [hunk 22]
```diff
@@ -0,0 +1,657 @@
...
+    properties.add(DiagnosticsProperty<FocusNode>('focusNode', focusNode, defaultValue: null));
+    properties.add(DiagnosticsProperty<BoxDecoration>('decoration', decoration));
+    properties.add(DiagnosticsProperty<EdgeInsetsGeometry>('padding', padding));
+    properties.add(StringProperty('placeholder', placeholder));
+    if (prefix != null) {
+      properties.add(DiagnosticsProperty<OverlayVisibilityMode>('prefix', prefixMode));
+    }
+    if (suffix != null) {
+      properties.add(DiagnosticsProperty<OverlayVisibilityMode>('suffix', suffixMode));
+    }
```
Comment (Hixie) [inline packages/flutter/lib/src/cupertino/text_field.dart]: we should always add these, just add them with defaultValue: null

File: packages/flutter/lib/src/cupertino/text_field.dart [hunk 23]
```diff
@@ -0,0 +1,657 @@
...
+
+  @override
+  void didUpdateWidget(CupertinoTextField oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    if (widget.controller == null && oldWidget.controller != null) {
+      _controller = TextEditingController.fromValue(oldWidget.controller.value);
+      _controller.addListener(updateKeepAlive);
+    }
+    else if (widget.controller != null && oldWidget.controller == null)
+      _controller = null;
```
Comment (Hixie) [inline packages/flutter/lib/src/cupertino/text_field.dart]: nits: `else` should be on the same line as the `}`.
If one block has `{}`, all should.

File: packages/flutter/lib/src/cupertino/text_field.dart [hunk 24]
```diff
@@ -0,0 +1,657 @@
...
+    _renderEditable.handleTapDown(details);
+  }
+
+  void _handleTap() {
+    _renderEditable.handleTap();
+    _requestKeyboard();
+  }
+
+  void _handleTapCancel() {
+  }
```
Comment (Hixie) [inline packages/flutter/lib/src/cupertino/text_field.dart]: ?

File: packages/flutter/lib/src/cupertino/text_field.dart [hunk 25]
```diff
@@ -0,0 +1,657 @@
...
+    _renderEditable.handleLongPress();
+  }
+
+  @override
+  bool get wantKeepAlive => _controller?.text?.isNotEmpty == true;
+
+  @override
+  void deactivate() {
+    super.deactivate();
+  }
```
Comment (Hixie) [inline packages/flutter/lib/src/cupertino/text_field.dart]: ?

File: packages/flutter/lib/src/cupertino/text_field.dart [hunk 26]
```diff
@@ -0,0 +1,657 @@
...
+      case OverlayVisibilityMode.never:
+        return false;
+      case OverlayVisibilityMode.always:
+        return true;
+      case OverlayVisibilityMode.editing:
+        return hasText;
+      case OverlayVisibilityMode.notEditing:
+        return !hasText;
+    }
+    assert(false);
```
Comment (Hixie) [inline packages/flutter/lib/src/cupertino/text_field.dart]: assert that attachment is non-null before the switch, then you can remove this assert.

File: packages/flutter/lib/src/cupertino/text_field.dart [hunk 27]
```diff
@@ -0,0 +1,657 @@
...
+        final List<Widget> rowChildren = <Widget>[];
+
+        // Insert a prefix at the front if the prefix visibility mode matches
+        // the current text state.
+        if (widget.prefix != null &&
+            _shouldShowAttachment(
+              attachment: widget.prefixMode,
+              hasText: text.text.isNotEmpty,
+            )
+        ) {
```
Comment (Hixie) [inline packages/flutter/lib/src/cupertino/text_field.dart]: nit: this wrapping's a bit weird. :-)

File: packages/flutter/lib/src/cupertino/text_field.dart [hunk 28]
```diff
@@ -0,0 +1,657 @@
...
+          rowChildren.add(widget.suffix);
+        // Otherwise, try to show a clear button if its visibility mode matches.
+        } else if (
+          _shouldShowAttachment(
+            attachment: widget.clearButtonMode,
+            hasText: text.text.isNotEmpty,
+          )
+        ) {
```
Comment (Hixie) [inline packages/flutter/lib/src/cupertino/text_field.dart]: i guess technically this is flutter-style wrapping but it's very confusing to look at.

File: packages/flutter/lib/src/cupertino/text_field.dart (hunk 20 above)
Comment (xster) [reply packages/flutter/lib/src/cupertino/text_field.dart]: Thanks for review. Done

File: packages/flutter/lib/src/cupertino/text_field.dart (hunk 21 above)
Comment (xster) [reply packages/flutter/lib/src/cupertino/text_field.dart]: done

File: packages/flutter/lib/src/cupertino/text_field.dart (hunk 22 above)
Comment (xster) [reply packages/flutter/lib/src/cupertino/text_field.dart]: done

File: packages/flutter/lib/src/cupertino/text_field.dart (hunk 23 above)
Comment (xster) [reply packages/flutter/lib/src/cupertino/text_field.dart]: pasta copy

File: packages/flutter/lib/src/cupertino/text_field.dart (hunk 24 above)
Comment (xster) [reply packages/flutter/lib/src/cupertino/text_field.dart]: doh, thanks

File: packages/flutter/lib/src/cupertino/text_field.dart (hunk 25 above)
Comment (xster) [reply packages/flutter/lib/src/cupertino/text_field.dart]: thanks

File: packages/flutter/lib/src/cupertino/text_field.dart (hunk 26 above)
Comment (xster) [reply packages/flutter/lib/src/cupertino/text_field.dart]: not sure if I understood correctly. This is just to make sure adding new enum values in the future fails visibly

File: packages/flutter/lib/src/cupertino/text_field.dart (hunk 27 above)
Comment (xster) [reply packages/flutter/lib/src/cupertino/text_field.dart]: refactored

File: packages/flutter/lib/src/cupertino/text_field.dart (hunk 28 above)
Comment (xster) [reply packages/flutter/lib/src/cupertino/text_field.dart]: same

File: packages/flutter/lib/src/cupertino/text_field.dart (hunk 16 above)
Comment (xster) [reply packages/flutter/lib/src/cupertino/text_field.dart]: Never mind. Did what you suggested

**Comment by xster** [comment]
Gentle ping. Any more comments?

**Comment by matthew-carroll** [comment]
LGTM!
